source/common/upstream/cluster_manager_impl.cc

```diff
@@ -487,9 +487,21 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
 
   if (existing_active_cluster != active_clusters_.end() ||
       existing_warming_cluster != warming_clusters_.end()) {
-    // The following init manager remove call is a NOP in the case we are already initialized. It's
-    // just kept here to avoid additional logic.
-    init_helper_.removeCluster(*existing_active_cluster->second->cluster_);
+    if (existing_active_cluster != active_clusters_.end()) {
+      // The following init manager remove call is a NOP in the case we are already initialized.
+      // It's just kept here to avoid additional logic.
+      init_helper_.removeCluster(*existing_active_cluster->second->cluster_);
+    } else {
+      // Validate that warming clusters are not added to the init_helper_.
+      for (const std::list<Cluster*>& cluster_list :
```

Comment:
Nit: should we #ifdef to avoid this iteration in non-debug mode?

Reply:
I wanted to avoid polluting the code with an `#ifdef` guard since the compiler optimizes away this for loop. I just confirmed by checking the disassembler output with a `-c opt` build.

Added a comment to clarify.